enhancements/fail-forward-upgrades.md

```diff
@@ -0,0 +1,252 @@
+- Allow operator authors to opt-out from failed upgrades when a specific operator version is deemed critical.
+
+### Non-Goals
+
+- Support operator rollbacks.
+- Allow OLM to automatically recover from CSVs in the PENDING phase.
+- Allow cluster admin to perform upgrades that Operator Authors deem unsafe.
+- Reattempt to install a failed installation.
+
+## Proposal
+
+OLM should be updated to allow cluster admins to recover from failed installations and upgrades. A failed installation/upgrade typically caused by one of two scenarios:
+1. **A Failed CSV:** The CSV is in the FAILED phase.
+2. **A Failed InstallPlan:** Usually occurs because a resource listed in the InstallPlan fails to be created or updated. An InstallPlan may fail independently of its CSV and may fail to create the CSV.
+
+This enhancement proposes that OLM be updated support "fail forward" upgrades by:
+- Providing cluster admins a means to opting into "fail forward" upgrades.
+- Allowing CSVs to move from the FAILED phase to the REPLACING phase.
+- Allowing OLM to calculate new installPlans for a set of instalables if:
+-- The InstallPlan referenced by a subscription is in the FAILED phase.
+-- The CatalogSource has been updated to include a new upgrade for one or more CSVs in the namespace.
+- Introducing a [Bundle Property](https://olm.operatorframework.io/docs/reference/file-based-catalogs/#properties) that allows operator authors to prevent a version of an operator from "failing forward" when they include critical work, such as CR migrations.
+### User Stories
+
+- As a cluster admin, I want to allow operators in a specific namespace to "fail forward".
+- As a cluster admin, if an InstallPlan is in the FAILED state, I want OLM to create the next InstallPlan if the catalog has been updated and includes a new set of installables.
+- As a cluster admin, if a CSV is stuck in the FAILED state and I have enabled "fail forward" upgrades, I want to OLM allow the operator to upgrade to the next version.
+- As a catalog curator / operator author, I want to be able to prevent a specific version of the operator from "failing forward" as it might include a critical update such as a custom resource migration.
+
+### Implementation Details/Notes/Constraints [optional]
+
+#### Opting into Fail Forward Upgrades
+
+Since resolution is namespaced, the toggle for allowing "fail forward" upgrades should be namespaced as well to avoid having to add the extra burden on the resolver surrounding understanding which dependent operators can or cannot use the forced upgrade toggle.
+
+With this in mind, this enhancement proposes adding the `upgradeStrategy` field to an already watched namespaced resource, the OperatorGroup.
+
+```yaml
+apiVersion: operators.coreos.com/v1
+kind: OperatorGroup
+metadata:
+  name: foo
+  namespace: bar
+spec:
+  upgradeStrategy:
+     # Possible values include "default" or "failForward".
+    type: failForward
```

Comment:
Do we need a `type` subfield of `upgradeStrategy? Or can `upgradeStrategy` be a string?

Also, should we capitalize the first letter of the value (e.g. `FailForward`) to match other kube APIs (e.g. `imagePullPolicy: Always`, `dnsPolicy: ClusterFirst`, etc)

Lastly, I could imagine that we end up with other strategies that fail forward (e.g. fail forward, but only if we're failing forward to a bundle in the same z stream as the failed one). Is there a more precise name we could give this one to leave room for other potential fail forward policies?

Reply:
I do worry that when we announce this to operator authors, they're going to say "hold on a second! we have a bunch of existing bundle versions out there that can't be failed forward from! we can't go and rebuild all of those versions to include this property!"

What's our answer to that?

Reply:
One suggestion is:
1. Add a feature flag to OLM that is required to be set to enable any of this
2. That feature flag is tied to a support statement (leave it off to get support, turn it on to get fail forward behavior, but lose support)